# Regras de Neg√≥cio - Planos e Permiss√µes

## 1. Tipos de Planos

### 1.1. Plano FREE (Gratuito)

**Recursos:**
- ‚úÖ 1 p√°gina principal
- ‚úÖ At√© 3 sub-p√°ginas
- ‚úÖ Editor completo com markdown
- ‚úÖ Anexar imagens
- ‚úÖ Blocos de coment√°rios (info, warning, danger)
- ‚úÖ Auto-save
- ‚ùå Exportar PDF
- ‚ùå Colabora√ß√£o em tempo real
- ‚ùå Compartilhar p√°ginas

**Limita√ß√µes:**
```kotlin
data class FreePlanLimits(
    val maxMainPages: Int = 1,
    val maxSubPagesPerPage: Int = 3,
    val canExportPdf: Boolean = false,
    val canCollaborate: Boolean = false,
    val canShare: Boolean = false
)
```

### 1.2. Plano PRO (Profissional)

**Recursos:**
- ‚úÖ At√© 100 p√°ginas principais
- ‚úÖ At√© 10 sub-p√°ginas por p√°gina principal
- ‚úÖ Todos os recursos do FREE
- ‚úÖ **Exportar p√°ginas como PDF**
- ‚ùå Colabora√ß√£o em tempo real
- ‚ùå Compartilhar p√°ginas

**Limita√ß√µes:**
```kotlin
data class ProPlanLimits(
    val maxMainPages: Int = 100,
    val maxSubPagesPerPage: Int = 10,
    val canExportPdf: Boolean = true,
    val canCollaborate: Boolean = false,
    val canShare: Boolean = false
)
```

### 1.3. Plano PREMIUM (Ilimitado)

**Recursos:**
- ‚úÖ P√°ginas principais ilimitadas
- ‚úÖ Sub-p√°ginas ilimitadas
- ‚úÖ Todos os recursos do PRO
- ‚úÖ **Colabora√ß√£o em tempo real**
- ‚úÖ **Compartilhar p√°ginas com outros usu√°rios**
- ‚úÖ **Edi√ß√£o simult√¢nea**
- ‚úÖ Hist√≥rico de vers√µes (futuro)
- ‚úÖ Suporte priorit√°rio (futuro)

**Limita√ß√µes:**
```kotlin
data class PremiumPlanLimits(
    val maxMainPages: Int = Int.MAX_VALUE, // Ilimitado
    val maxSubPagesPerPage: Int = Int.MAX_VALUE,
    val canExportPdf: Boolean = true,
    val canCollaborate: Boolean = true,
    val canShare: Boolean = true
)
```

---

## 2. Enum de Planos

```kotlin
enum class PlanType {
    FREE,
    PRO,
    PREMIUM;
    
    fun getLimits(): PlanLimits {
        return when(this) {
            FREE -> FreePlanLimits()
            PRO -> ProPlanLimits()
            PREMIUM -> PremiumPlanLimits()
        }
    }
}
```

---

## 3. Valida√ß√£o de Limites

### 3.1. Cria√ß√£o de P√°gina Principal

**Endpoint**: POST /api/documents

**Valida√ß√µes:**
1. Verificar plano do usu√°rio
2. Contar p√°ginas principais existentes do usu√°rio
3. Comparar com limite do plano
4. Se exceder, retornar erro HTTP 403

**Exemplo:**
```kotlin
fun validateCanCreateMainPage(userId: UUID): Boolean {
    val user = userRepository.findById(userId)
    val currentCount = documentRepository.countMainPagesByUserId(userId)
    val limit = user.plan.getLimits().maxMainPages
    
    if (currentCount >= limit) {
        throw PlanLimitExceededException(
            "Voc√™ atingiu o limite de ${limit} p√°ginas principais do plano ${user.plan}. " +
            "Fa√ßa upgrade para criar mais p√°ginas."
        )
    }
    return true
}
```

### 3.2. Cria√ß√£o de Sub-p√°gina

**Valida√ß√µes:**
1. Verificar plano do usu√°rio
2. Contar sub-p√°ginas da p√°gina pai
3. Comparar com limite do plano
4. Se exceder, retornar erro HTTP 403

**Exemplo:**
```kotlin
fun validateCanCreateSubPage(userId: UUID, parentPageId: UUID): Boolean {
    val user = userRepository.findById(userId)
    val currentCount = documentRepository.countSubPagesByParentId(parentPageId)
    val limit = user.plan.getLimits().maxSubPagesPerPage
    
    if (currentCount >= limit) {
        throw PlanLimitExceededException(
            "Voc√™ atingiu o limite de ${limit} sub-p√°ginas por p√°gina do plano ${user.plan}. " +
            "Fa√ßa upgrade para criar mais sub-p√°ginas."
        )
    }
    return true
}
```

### 3.3. Exportar PDF

**Endpoint**: GET /api/documents/:id/export/pdf

**Valida√ß√µes:**
1. Verificar plano do usu√°rio
2. Se FREE, negar acesso
3. Se PRO ou PREMIUM, permitir

```kotlin
@PreAuthorize("hasAuthority('EXPORT_PDF')")
fun exportToPdf(documentId: UUID): ByteArray {
    // Gera PDF
}
```

**Authorities por Plano:**
```kotlin
fun getAuthoritiesByPlan(plan: PlanType): List<String> {
    return when(plan) {
        FREE -> listOf("READ", "WRITE")
        PRO -> listOf("READ", "WRITE", "EXPORT_PDF")
        PREMIUM -> listOf("READ", "WRITE", "EXPORT_PDF", "COLLABORATE", "SHARE")
    }
}
```

### 3.4. Compartilhar P√°gina

**Endpoint**: POST /api/documents/:id/share

**Valida√ß√µes:**
1. Verificar se plano √© PREMIUM
2. Se n√£o for, retornar erro HTTP 403

```kotlin
@PreAuthorize("hasAuthority('SHARE')")
fun shareDocument(documentId: UUID, shareWith: List<String>): ShareResponse {
    // Compartilha documento
}
```

### 3.5. Colabora√ß√£o em Tempo Real

**WebSocket Connect:**

**Valida√ß√µes:**
1. Ao conectar no WebSocket de um documento
2. Verificar plano do DONO do documento
3. Se n√£o for PREMIUM, rejeitar conex√£o

```kotlin
@MessageMapping("/document/{id}/join")
fun joinDocument(@DestinationVariable id: UUID, principal: Principal) {
    val document = documentRepository.findById(id)
    val owner = userRepository.findById(document.ownerId)
    
    if (owner.plan != PlanType.PREMIUM) {
        throw AccessDeniedException(
            "Colabora√ß√£o em tempo real requer plano Premium"
        )
    }
    
    // Permite entrar na sess√£o
}
```

---

## 4. Upgrade e Downgrade de Plano

### 4.1. Upgrade de Plano

**Endpoint**: POST /api/users/me/upgrade

**Payload:**
```json
{
  "newPlan": "PRO",
  "paymentMethod": "credit_card",
  "paymentDetails": {}
}
```

**Fluxo:**
1. Validar novo plano (deve ser superior ao atual)
2. Processar pagamento (integra√ß√£o futura)
3. Atualizar plano do usu√°rio
4. Conceder novas permiss√µes imediatamente
5. Enviar email de confirma√ß√£o
6. Retornar sucesso

**Regras:**
- FREE ‚Üí PRO: Permitido
- FREE ‚Üí PREMIUM: Permitido
- PRO ‚Üí PREMIUM: Permitido
- Upgrade √© imediato ap√≥s pagamento

### 4.2. Downgrade de Plano

**Endpoint**: POST /api/users/me/downgrade

**Payload:**
```json
{
  "newPlan": "FREE",
  "reason": "TOO_EXPENSIVE"
}
```

**Fluxo:**
1. Validar novo plano (deve ser inferior ao atual)
2. **Verificar se h√° conte√∫do que excede limites**
3. Se exceder, exigir que usu√°rio delete/arquive conte√∫do antes
4. Cancelar assinatura
5. Agendar downgrade para fim do per√≠odo pago
6. Enviar email de confirma√ß√£o

**Valida√ß√µes de Conte√∫do:**

```kotlin
fun validateCanDowngrade(userId: UUID, newPlan: PlanType): DowngradeValidation {
    val user = userRepository.findById(userId)
    val newLimits = newPlan.getLimits()
    
    val mainPagesCount = documentRepository.countMainPagesByUserId(userId)
    val subPagesExceeded = documentRepository.findPagesExceedingSubPageLimit(
        userId, 
        newLimits.maxSubPagesPerPage
    )
    
    val validation = DowngradeValidation(
        canDowngrade = true,
        issues = mutableListOf()
    )
    
    // Verifica limite de p√°ginas principais
    if (mainPagesCount > newLimits.maxMainPages) {
        validation.canDowngrade = false
        validation.issues.add(
            "Voc√™ tem ${mainPagesCount} p√°ginas principais, " +
            "mas o plano ${newPlan} permite apenas ${newLimits.maxMainPages}. " +
            "Delete ${mainPagesCount - newLimits.maxMainPages} p√°ginas antes de fazer downgrade."
        )
    }
    
    // Verifica sub-p√°ginas
    if (subPagesExceeded.isNotEmpty()) {
        validation.canDowngrade = false
        validation.issues.add(
            "As seguintes p√°ginas excedem o limite de sub-p√°ginas: " +
            subPagesExceeded.joinToString { it.title }
        )
    }
    
    return validation
}
```

**Resposta se N√£o Puder Fazer Downgrade:**

```json
{
  "canDowngrade": false,
  "issues": [
    "Voc√™ tem 5 p√°ginas principais, mas o plano FREE permite apenas 1. Delete 4 p√°ginas antes de fazer downgrade.",
    "A p√°gina 'Projeto X' tem 8 sub-p√°ginas, mas o plano FREE permite apenas 3."
  ]
}
```

### 4.3. Comportamento Ap√≥s Downgrade

**Se downgrade para FREE:**
1. Desabilitar exporta√ß√£o de PDF
2. Desabilitar compartilhamento (manter compartilhamentos existentes como read-only)
3. Desabilitar colabora√ß√£o em tempo real
4. Manter conte√∫do existente (n√£o deletar)
5. Impedir cria√ß√£o de novo conte√∫do al√©m dos limites

---

## 5. Mensagens de Limite Atingido

### 5.1. No Frontend

Quando usu√°rio tenta criar p√°gina e atingiu limite:

**Modal:**
```
Limite do Plano Atingido üö´

Voc√™ atingiu o limite de 1 p√°gina principal do plano FREE.

Para criar mais p√°ginas, fa√ßa upgrade para:

üì¶ Plano PRO - R$ 29,90/m√™s
- At√© 100 p√°ginas principais
- At√© 10 sub-p√°ginas por p√°gina
- Exportar PDF

üíé Plano PREMIUM - R$ 59,90/m√™s
- P√°ginas ilimitadas
- Sub-p√°ginas ilimitadas
- Exportar PDF
- Colabora√ß√£o em tempo real

[Ver Compara√ß√£o de Planos]  [Fazer Upgrade]
```

### 5.2. Banner de Aviso

Quando usu√°rio est√° pr√≥ximo do limite (80%):

**Banner:**
```
‚ö†Ô∏è Voc√™ est√° usando 4 de 5 p√°ginas principais do seu plano PRO.
[Fazer Upgrade para Premium]
```

### 5.3. P√°gina de Compara√ß√£o

Exibir tabela comparativa de planos:

| Recurso | FREE | PRO | PREMIUM |
|---------|------|-----|---------|
| P√°ginas principais | 1 | 100 | Ilimitadas |
| Sub-p√°ginas por p√°gina | 3 | 10 | Ilimitadas |
| Editor completo | ‚úÖ | ‚úÖ | ‚úÖ |
| Imagens | ‚úÖ | ‚úÖ | ‚úÖ |
| Auto-save | ‚úÖ | ‚úÖ | ‚úÖ |
| Exportar PDF | ‚ùå | ‚úÖ | ‚úÖ |
| Compartilhar | ‚ùå | ‚ùå | ‚úÖ |
| Colabora√ß√£o em tempo real | ‚ùå | ‚ùå | ‚úÖ |
| **Pre√ßo** | **Gr√°tis** | **R$ 29,90/m√™s** | **R$ 59,90/m√™s** |

---

## 6. Permiss√µes Detalhadas

### 6.1. Matriz de Permiss√µes

```kotlin
enum class Permission {
    // Documentos
    CREATE_DOCUMENT,
    READ_DOCUMENT,
    UPDATE_DOCUMENT,
    DELETE_DOCUMENT,
    
    // Sub-p√°ginas
    CREATE_SUB_PAGE,
    
    // Exporta√ß√£o
    EXPORT_PDF,
    
    // Compartilhamento
    SHARE_DOCUMENT,
    VIEW_SHARED_DOCUMENT,
    EDIT_SHARED_DOCUMENT,
    
    // Colabora√ß√£o
    COLLABORATE_REALTIME,
    
    // Admin (futuro)
    MANAGE_USERS,
    VIEW_ANALYTICS
}

fun getPermissionsByPlan(plan: PlanType): Set<Permission> {
    val base = setOf(
        Permission.CREATE_DOCUMENT,
        Permission.READ_DOCUMENT,
        Permission.UPDATE_DOCUMENT,
        Permission.DELETE_DOCUMENT,
        Permission.CREATE_SUB_PAGE
    )
    
    return when(plan) {
        FREE -> base
        PRO -> base + setOf(Permission.EXPORT_PDF)
        PREMIUM -> base + setOf(
            Permission.EXPORT_PDF,
            Permission.SHARE_DOCUMENT,
            Permission.VIEW_SHARED_DOCUMENT,
            Permission.EDIT_SHARED_DOCUMENT,
            Permission.COLLABORATE_REALTIME
        )
    }
}
```

### 6.2. Verifica√ß√£o de Permiss√£o

**Annotation Customizada:**
```kotlin
@Target(AnnotationTarget.FUNCTION)
@Retention(AnnotationRetention.RUNTIME)
annotation class RequiresPlan(val minPlan: PlanType)

// Uso:
@RequiresPlan(PlanType.PRO)
fun exportToPdf(documentId: UUID): ByteArray {
    // ...
}
```

**Aspect para Valida√ß√£o:**
```kotlin
@Aspect
class PlanPermissionAspect {
    
    @Before("@annotation(requiresPlan)")
    fun checkPlanPermission(joinPoint: JoinPoint, requiresPlan: RequiresPlan) {
        val user = SecurityContextHolder.getContext().authentication.principal as User
        
        if (user.plan < requiresPlan.minPlan) {
            throw InsufficientPlanException(
                "Este recurso requer plano ${requiresPlan.minPlan} ou superior"
            )
        }
    }
}
```

---

## 7. Estat√≠sticas e Uso

### 7.1. Endpoint de Estat√≠sticas

**Endpoint**: GET /api/users/me/usage

**Resposta:**
```json
{
  "plan": "PRO",
  "usage": {
    "mainPages": {
      "current": 45,
      "limit": 100,
      "percentage": 45
    },
    "subPages": {
      "highest": 8,
      "limit": 10,
      "percentage": 80,
      "pageWithMostSubPages": "Projeto X"
    },
    "storage": {
      "used": "125 MB",
      "limit": "1 GB",
      "percentage": 12.5
    }
  },
  "recommendations": [
    "Voc√™ est√° usando 80% do limite de sub-p√°ginas na p√°gina 'Projeto X'",
    "Considere fazer upgrade para Premium para p√°ginas ilimitadas"
  ]
}
```

### 7.2. Notifica√ß√µes de Uso

**Enviar email quando:**
- Usu√°rio atinge 80% de qualquer limite
- Usu√°rio atinge 100% de qualquer limite
- Downgrade est√° agendado (7 dias antes)

---

## 8. Trial Premium

### 8.1. Per√≠odo de Teste

**Oferecer 14 dias de trial Premium:**
- Dispon√≠vel apenas para usu√°rios FREE ou PRO
- Apenas 1 trial por usu√°rio (lifetime)
- Acesso total aos recursos Premium
- N√£o requer cart√£o de cr√©dito
- Ap√≥s expirar, retorna ao plano anterior

**Endpoint**: POST /api/users/me/start-trial

**Valida√ß√µes:**
1. Usu√°rio nunca usou trial antes?
2. Plano atual √© FREE ou PRO?
3. Criar registro de trial com data de expira√ß√£o

```kotlin
data class Trial(
    val userId: UUID,
    val startDate: LocalDateTime,
    val endDate: LocalDateTime,
    val originalPlan: PlanType,
    val status: TrialStatus  // ACTIVE, EXPIRED, CONVERTED
)
```

### 8.2. Expira√ß√£o do Trial

**Job agendado di√°rio:**
1. Buscar trials com endDate < hoje
2. Reverter usu√°rios para plano original
3. Enviar email "Seu trial expirou"
4. Oferecer desconto para converter

---

## 9. C√≥digos de Desconto

### 9.1. Cupons Promocionais

```kotlin
data class DiscountCoupon(
    val code: String,              // "BEMVINDO2024"
    val discountType: DiscountType, // PERCENTAGE, FIXED_AMOUNT
    val discountValue: BigDecimal,  // 20 (20% ou R$ 20)
    val validFrom: LocalDateTime,
    val validUntil: LocalDateTime,
    val maxUses: Int?,              // null = ilimitado
    val currentUses: Int,
    val applicablePlans: List<PlanType>,
    val firstPurchaseOnly: Boolean
)

enum class DiscountType {
    PERCENTAGE,      // 20% off
    FIXED_AMOUNT,    // R$ 10 off
    FIRST_MONTH_FREE // Primeiro m√™s gr√°tis
}
```

### 9.2. Aplicar Cupom

**Endpoint**: POST /api/users/me/apply-coupon

**Payload:**
```json
{
  "code": "BEMVINDO2024",
  "plan": "PRO"
}
```

**Valida√ß√µes:**
1. Cupom existe?
2. Est√° dentro da validade?
3. N√£o excedeu m√°ximo de usos?
4. Aplic√°vel ao plano escolhido?
5. Se firstPurchaseOnly, usu√°rio nunca pagou antes?

---

## 10. Mensagens de Erro de Plano

```json
{
  "error": {
    "code": "PLAN_LIMIT_EXCEEDED",
    "message": "Voc√™ atingiu o limite de p√°ginas principais do plano FREE",
    "details": {
      "currentPlan": "FREE",
      "limit": 1,
      "current": 1,
      "requiredPlan": "PRO"
    },
    "actions": [
      {
        "label": "Fazer Upgrade",
        "url": "/pricing"
      },
      {
        "label": "Ver Compara√ß√£o",
        "url": "/plans/compare"
      }
    ]
  }
}
```

**C√≥digos de Erro:**
- `PLAN_LIMIT_EXCEEDED` - Limite atingido
- `INSUFFICIENT_PLAN` - Plano insuficiente para recurso
- `TRIAL_ALREADY_USED` - Trial j√° utilizado
- `INVALID_COUPON` - Cupom inv√°lido
- `CANNOT_DOWNGRADE` - N√£o pode fazer downgrade
