# Regras de NegÃ³cio - Documentos e Editor

## 1. Estrutura de Documentos

### 1.1. Hierarquia

```
Workspace (UsuÃ¡rio)
â”œâ”€â”€ PÃ¡gina Principal 1
â”‚   â”œâ”€â”€ Sub-pÃ¡gina 1.1
â”‚   â”œâ”€â”€ Sub-pÃ¡gina 1.2
â”‚   â””â”€â”€ Sub-pÃ¡gina 1.3
â”œâ”€â”€ PÃ¡gina Principal 2
â”‚   â”œâ”€â”€ Sub-pÃ¡gina 2.1
â”‚   â”‚   â””â”€â”€ âŒ NÃ£o permite sub-sub-pÃ¡ginas
â”‚   â””â”€â”€ Sub-pÃ¡gina 2.2
â””â”€â”€ PÃ¡gina Principal 3
```

**Regras:**
- MÃ¡ximo 2 nÃ­veis de hierarquia (pÃ¡gina principal + sub-pÃ¡ginas)
- NÃ£o permite sub-sub-pÃ¡ginas (netos)
- Cada sub-pÃ¡gina pertence a apenas uma pÃ¡gina pai

### 1.2. Model de Documento

```kotlin
@Entity
@Table(name = "documents")
data class Document(
    @Id
    val id: UUID = UUID.randomUUID(),
    
    @Column(nullable = false)
    val title: String,
    
    @Column(columnDefinition = "TEXT")
    val content: String = "",
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "owner_id", nullable = false)
    val owner: User,
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent_id")
    val parent: Document? = null,  // null = pÃ¡gina principal
    
    @OneToMany(mappedBy = "parent", cascade = [CascadeType.ALL])
    val subPages: MutableList<Document> = mutableListOf(),
    
    @Column(nullable = false)
    val icon: String? = "ğŸ“„",  // Emoji como Ã­cone
    
    @Column(nullable = false)
    val coverImage: String? = null,  // URL da imagem de capa
    
    @Enumerated(EnumType.STRING)
    val status: DocumentStatus = DocumentStatus.DRAFT,
    
    @Column(nullable = false)
    val isPublic: Boolean = false,
    
    @Column(nullable = false)
    val isFavorite: Boolean = false,
    
    @Column(nullable = false)
    val isArchived: Boolean = false,
    
    @CreatedDate
    val createdAt: LocalDateTime = LocalDateTime.now(),
    
    @LastModifiedDate
    val updatedAt: LocalDateTime = LocalDateTime.now(),
    
    @Column
    val lastEditedBy: UUID? = null,
    
    @Column
    val deletedAt: LocalDateTime? = null
)

enum class DocumentStatus {
    DRAFT,      // Rascunho
    PUBLISHED,  // Publicado
    ARCHIVED    // Arquivado
}
```

---

## 2. CRUD de Documentos

### 2.1. Criar Documento

**Endpoint**: POST /api/documents

**Payload:**
```json
{
  "title": "Minha Nova PÃ¡gina",
  "parentId": null,  // null = pÃ¡gina principal
  "icon": "ğŸ“",
  "coverImage": null
}
```

**ValidaÃ§Ãµes:**
1. TÃ­tulo Ã© obrigatÃ³rio (mÃ­n 1, mÃ¡x 255 caracteres)
2. Se parentId for fornecido:
   - PÃ¡gina pai deve existir
   - PÃ¡gina pai deve pertencer ao usuÃ¡rio ou ser compartilhada
   - PÃ¡gina pai nÃ£o pode ser sub-pÃ¡gina (max 2 nÃ­veis)
3. Validar limites do plano (ver regras de planos)
4. Se criar pÃ¡gina principal, verificar limite de pÃ¡ginas principais
5. Se criar sub-pÃ¡gina, verificar limite de sub-pÃ¡ginas da pÃ¡gina pai

**Resposta Sucesso (HTTP 201):**
```json
{
  "id": "uuid",
  "title": "Minha Nova PÃ¡gina",
  "content": "",
  "parentId": null,
  "icon": "ğŸ“",
  "coverImage": null,
  "status": "DRAFT",
  "isPublic": false,
  "isFavorite": false,
  "createdAt": "2024-01-01T00:00:00Z",
  "updatedAt": "2024-01-01T00:00:00Z"
}
```

### 2.2. Listar Documentos

**Endpoint**: GET /api/documents

**Query Params:**
- `parentId` (UUID, opcional) - Listar sub-pÃ¡ginas de uma pÃ¡gina
- `status` (DRAFT|PUBLISHED|ARCHIVED, opcional)
- `favorite` (boolean, opcional)
- `archived` (boolean, opcional)
- `page` (number, default: 0)
- `size` (number, default: 20)
- `sort` (string, default: "updatedAt,desc")

**Exemplos:**
```
GET /api/documents
â†’ Lista todas pÃ¡ginas principais do usuÃ¡rio

GET /api/documents?parentId=uuid
â†’ Lista sub-pÃ¡ginas da pÃ¡gina especificada

GET /api/documents?favorite=true
â†’ Lista apenas favoritos

GET /api/documents?archived=true
â†’ Lista apenas arquivados

GET /api/documents?status=PUBLISHED&sort=title,asc
â†’ Lista publicados ordenados por tÃ­tulo
```

**Resposta:**
```json
{
  "content": [
    {
      "id": "uuid",
      "title": "Projeto X",
      "icon": "ğŸš€",
      "status": "PUBLISHED",
      "isFavorite": true,
      "subPagesCount": 5,
      "updatedAt": "2024-01-01T00:00:00Z"
    }
  ],
  "page": 0,
  "size": 20,
  "totalElements": 45,
  "totalPages": 3
}
```

### 2.3. Obter Documento

**Endpoint**: GET /api/documents/:id

**ValidaÃ§Ãµes:**
1. Documento existe?
2. UsuÃ¡rio tem permissÃ£o de leitura?
   - Ã‰ o dono?
   - Documento foi compartilhado com ele?
   - Documento Ã© pÃºblico?

**Resposta:**
```json
{
  "id": "uuid",
  "title": "Minha PÃ¡gina",
  "content": "# TÃ­tulo\n\nConteÃºdo...",
  "parentId": null,
  "parent": null,
  "subPages": [
    {
      "id": "uuid",
      "title": "Sub-pÃ¡gina 1",
      "icon": "ğŸ“„"
    }
  ],
  "icon": "ğŸ“",
  "coverImage": "https://...",
  "status": "PUBLISHED",
  "isPublic": false,
  "isFavorite": false,
  "owner": {
    "id": "uuid",
    "name": "JoÃ£o Silva",
    "username": "joao",
    "avatar": "https://..."
  },
  "createdAt": "2024-01-01T00:00:00Z",
  "updatedAt": "2024-01-01T12:30:00Z",
  "lastEditedBy": {
    "id": "uuid",
    "name": "Maria Santos",
    "username": "maria"
  }
}
```

### 2.4. Atualizar Documento

**Endpoint**: PATCH /api/documents/:id

**Payload:**
```json
{
  "title": "Novo TÃ­tulo",
  "content": "# ConteÃºdo atualizado",
  "icon": "ğŸ¯",
  "coverImage": "https://...",
  "status": "PUBLISHED"
}
```

**ValidaÃ§Ãµes:**
1. Documento existe?
2. UsuÃ¡rio tem permissÃ£o de escrita?
3. Se mudar status para PUBLISHED, validar que tem conteÃºdo

**Comportamento:**
- Atualiza apenas campos fornecidos (merge)
- Atualiza `updatedAt`
- Atualiza `lastEditedBy` com ID do usuÃ¡rio atual
- Se houver colaboradores online, emite evento WebSocket

### 2.5. Deletar Documento

**Endpoint**: DELETE /api/documents/:id

**Comportamento: Soft Delete**
1. NÃ£o deleta fisicamente
2. Atualiza `deletedAt` com timestamp atual
3. Queries normais nÃ£o retornam documentos deletados
4. MantÃ©m por 30 dias para recuperaÃ§Ã£o
5. Job agendado limpa definitivamente apÃ³s 30 dias

**ValidaÃ§Ãµes:**
1. Documento existe?
2. UsuÃ¡rio Ã© o dono?
3. Se deletar pÃ¡gina principal, todos sub-pÃ¡ginas sÃ£o deletadas em cascata

**Resposta:**
```json
{
  "message": "Documento movido para lixeira",
  "deletedAt": "2024-01-01T12:00:00Z",
  "permanentDeletionAt": "2024-01-31T12:00:00Z"
}
```

### 2.6. Restaurar da Lixeira

**Endpoint**: POST /api/documents/:id/restore

**ValidaÃ§Ãµes:**
1. Documento estÃ¡ deletado?
2. UsuÃ¡rio Ã© o dono?
3. Ainda estÃ¡ dentro dos 30 dias?
4. Restaurar respeitando limites do plano atual

---

## 3. Editor de Texto

### 3.1. Formato de ConteÃºdo

O conteÃºdo Ã© armazenado em **formato JSON** do Tiptap:

```json
{
  "type": "doc",
  "content": [
    {
      "type": "heading",
      "attrs": { "level": 1 },
      "content": [
        { "type": "text", "text": "TÃ­tulo Principal" }
      ]
    },
    {
      "type": "paragraph",
      "content": [
        { "type": "text", "text": "ParÃ¡grafo normal com " },
        { 
          "type": "text", 
          "marks": [{ "type": "bold" }],
          "text": "negrito" 
        }
      ]
    }
  ]
}
```

**Vantagens:**
- Estruturado e fÃ¡cil de parsear
- FÃ¡cil conversÃ£o para HTML
- Suporta extensÃµes customizadas
- Permite validaÃ§Ã£o de conteÃºdo

### 3.2. Atalhos Markdown Suportados

**TÃ­tulos:**
- `#` + espaÃ§o â†’ H1
- `##` + espaÃ§o â†’ H2
- `###` + espaÃ§o â†’ H3

**FormataÃ§Ã£o:**
- `*texto*` ou `_texto_` â†’ *itÃ¡lico*
- `**texto**` ou `__texto__` â†’ **negrito**
- `~~texto~~` â†’ ~~tachado~~
- `` `cÃ³digo` `` â†’ `cÃ³digo inline`

**Listas:**
- `-` + espaÃ§o â†’ Lista nÃ£o ordenada
- `*` + espaÃ§o â†’ Lista nÃ£o ordenada (alternativo)
- `1.` + espaÃ§o â†’ Lista ordenada
- `[ ]` + espaÃ§o â†’ Checkbox desmarcado
- `[x]` + espaÃ§o â†’ Checkbox marcado

**Blocos:**
- ` ``` ` + enter â†’ Bloco de cÃ³digo
- `>` + espaÃ§o â†’ CitaÃ§Ã£o (blockquote)
- `---` â†’ Linha horizontal

### 3.3. Comando Slash (/)

Digite `/` para abrir a paleta de comandos:

**Comandos DisponÃ­veis:**

**BÃ¡sicos:**
- `/heading1` ou `/h1` â†’ TÃ­tulo 1
- `/heading2` ou `/h2` â†’ TÃ­tulo 2
- `/heading3` ou `/h3` â†’ TÃ­tulo 3
- `/paragraph` â†’ ParÃ¡grafo normal
- `/bulletlist` â†’ Lista com bullets
- `/numberlist` â†’ Lista numerada
- `/checkbox` â†’ Lista de tarefas

**Blocos Especiais:**
- `/code` â†’ Bloco de cÃ³digo
- `/quote` â†’ CitaÃ§Ã£o

**Callouts:**
- `/info` â†’ Bloco de informaÃ§Ã£o (azul)
- `/warning` â†’ Bloco de aviso (amarelo)
- `/danger` â†’ Bloco de perigo (vermelho)
- `/success` â†’ Bloco de sucesso (verde)

**MÃ­dia:**
- `/image` â†’ Inserir imagem
- `/video` â†’ Inserir vÃ­deo (URL)
- `/divider` â†’ Linha divisÃ³ria

**AvanÃ§ado:**
- `/table` â†’ Tabela
- `/page` â†’ Criar sub-pÃ¡gina

### 3.4. Callout Blocks

**Tipos de Callouts:**

```typescript
type CalloutType = 'info' | 'warning' | 'danger' | 'success';

interface Callout {
  type: CalloutType;
  title?: string;
  content: string;
}
```

**RenderizaÃ§Ã£o:**

**Info (Azul):**
```
â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚ â„¹ï¸ InformaÃ§Ã£o                  â”‚
â”œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤
â”‚ ConteÃºdo informativo aqui...   â”‚
â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜
```

**Warning (Amarelo):**
```
â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚ âš ï¸ AtenÃ§Ã£o                     â”‚
â”œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤
â”‚ ConteÃºdo de aviso aqui...      â”‚
â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜
```

**Danger (Vermelho):**
```
â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚ ğŸš¨ Perigo                      â”‚
â”œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤
â”‚ ConteÃºdo crÃ­tico aqui...       â”‚
â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜
```

**Success (Verde):**
```
â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚ âœ… Sucesso                     â”‚
â”œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤
â”‚ Mensagem de sucesso aqui...    â”‚
â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜
```

**JSON Structure:**
```json
{
  "type": "callout",
  "attrs": {
    "type": "warning",
    "title": "AtenÃ§Ã£o"
  },
  "content": [
    {
      "type": "paragraph",
      "content": [
        { "type": "text", "text": "ConteÃºdo do callout..." }
      ]
    }
  ]
}
```

---

## 4. Auto-save

### 4.1. EstratÃ©gia de Salvamento

**Debounce de 1 segundo:**
1. UsuÃ¡rio digita no editor
2. Frontend espera 1 segundo sem mudanÃ§as
3. Envia PATCH /api/documents/:id com novo conteÃºdo
4. Backend salva no banco
5. Backend emite evento WebSocket (se colaboraÃ§Ã£o)

**Estados do Auto-save:**
- **Typing...** - UsuÃ¡rio estÃ¡ digitando
- **Saving...** - Enviando para backend
- **Saved** - Salvo com sucesso
- **Error** - Erro ao salvar (mostrar mensagem, tentar novamente)

### 4.2. Indicador Visual

**No header do documento:**
```
[Meu Documento] [âœï¸ Salvando...]
[Meu Documento] [âœ… Salvo Ã s 14:30]
[Meu Documento] [âŒ Erro ao salvar - Tentar novamente]
```

### 4.3. Payload de Auto-save

**Endpoint**: PATCH /api/documents/:id/content

**Payload:**
```json
{
  "content": "{...}",  // JSON do Tiptap
  "version": 123        // NÃºmero de versÃ£o para controle
}
```

**Versionamento:**
- Cada salvamento incrementa a versÃ£o
- Previne conflitos em salvamentos simultÃ¢neos
- Se versÃ£o recebida < versÃ£o atual, rejeitar (HTTP 409 Conflict)

### 4.4. Tratamento de Erros

**Erro de Rede:**
- Mostrar mensagem "Sem conexÃ£o. Suas alteraÃ§Ãµes serÃ£o salvas quando voltar online"
- Armazenar conteÃºdo no localStorage
- Tentar reenviar a cada 10 segundos
- Quando conexÃ£o voltar, salvar e limpar localStorage

**Erro de PermissÃ£o:**
- Documento foi descompartilhado
- Mostrar "VocÃª nÃ£o tem mais permissÃ£o para editar este documento"
- Desabilitar editor
- Oferecer opÃ§Ã£o de salvar cÃ³pia

**Erro de VersÃ£o (Conflito):**
- Outro usuÃ¡rio editou simultaneamente
- Mostrar diff das mudanÃ§as
- Permitir escolher qual versÃ£o manter
- Ou mesclar mudanÃ§as (merge)

---

## 5. Upload de Imagens

### 5.1. Endpoint de Upload

**Endpoint**: POST /api/documents/:id/images

**Content-Type**: multipart/form-data

**ValidaÃ§Ãµes:**
1. Documento existe?
2. UsuÃ¡rio tem permissÃ£o de escrita?
3. Tamanho mÃ¡ximo: 5MB
4. Formatos permitidos: jpg, jpeg, png, gif, webp
5. Validar que Ã© realmente uma imagem (magic bytes)

**Fluxo:**
1. Recebe arquivo
2. Valida tipo e tamanho
3. Gera nome Ãºnico (UUID + extensÃ£o)
4. Redimensiona para mÃ¡ximo 1920x1080 (manter proporÃ§Ã£o)
5. Gera thumbnail 300x300
6. Salva em storage (local ou S3)
7. Salva registro no banco
8. Retorna URLs

**Resposta (HTTP 201):**
```json
{
  "id": "uuid",
  "url": "https://storage.com/images/uuid.jpg",
  "thumbnail": "https://storage.com/images/uuid-thumb.jpg",
  "filename": "minha-imagem.jpg",
  "size": 1024000,
  "width": 1920,
  "height": 1080,
  "uploadedAt": "2024-01-01T12:00:00Z"
}
```

### 5.2. Model de Imagem

```kotlin
@Entity
@Table(name = "document_images")
data class DocumentImage(
    @Id
    val id: UUID = UUID.randomUUID(),
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "document_id", nullable = false)
    val document: Document,
    
    @Column(nullable = false)
    val url: String,
    
    @Column(nullable = false)
    val thumbnailUrl: String,
    
    @Column(nullable = false)
    val filename: String,
    
    @Column(nullable = false)
    val size: Long,
    
    @Column
    val width: Int?,
    
    @Column
    val height: Int?,
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "uploaded_by", nullable = false)
    val uploadedBy: User,
    
    @CreatedDate
    val uploadedAt: LocalDateTime = LocalDateTime.now()
)
```

### 5.3. Deletar Imagem

**Endpoint**: DELETE /api/documents/:docId/images/:imageId

**Comportamento:**
1. Remove do storage
2. Remove registro do banco
3. Se imagem estiver no conteÃºdo do documento, mantÃ©m (broken link)
4. Frontend pode detectar broken links e removÃª-los

### 5.4. Limites de Storage (Futuro)

**Por Plano:**
- FREE: 100 MB
- PRO: 1 GB
- PREMIUM: 10 GB

---

## 6. Ãcones e Capas

### 6.1. Ãcone da PÃ¡gina

**Tipos Suportados:**
- **Emoji**: ğŸ“, ğŸš€, ğŸ’¡, etc.
- **URL de Imagem**: https://...
- **Iniciais**: Gerar automaticamente das iniciais do tÃ­tulo

**Endpoint**: PATCH /api/documents/:id/icon

**Payload:**
```json
{
  "icon": "ğŸ¯"
}
```

### 6.2. Imagem de Capa

**Endpoint**: POST /api/documents/:id/cover

**Content-Type**: multipart/form-data

**ValidaÃ§Ãµes:**
- Mesmas validaÃ§Ãµes de upload de imagem
- Redimensionar para 1920x400 (banner)

**Endpoint**: DELETE /api/documents/:id/cover

Remove a imagem de capa.

---

## 7. Favoritos

### 7.1. Adicionar aos Favoritos

**Endpoint**: POST /api/documents/:id/favorite

**Comportamento:**
- Atualiza `isFavorite` para `true`
- Adiciona ao topo da lista de favoritos

### 7.2. Remover dos Favoritos

**Endpoint**: DELETE /api/documents/:id/favorite

**Comportamento:**
- Atualiza `isFavorite` para `false`

### 7.3. Listar Favoritos

**Endpoint**: GET /api/documents?favorite=true

Retorna apenas documentos favoritados.

---

## 8. Arquivamento

### 8.1. Arquivar Documento

**Endpoint**: POST /api/documents/:id/archive

**Comportamento:**
- Atualiza `isArchived` para `true`
- Remove da lista principal
- DisponÃ­vel na seÃ§Ã£o "Arquivados"

### 8.2. Desarquivar

**Endpoint**: POST /api/documents/:id/unarchive

**Comportamento:**
- Atualiza `isArchived` para `false`
- Retorna Ã  lista principal

### 8.3. Listar Arquivados

**Endpoint**: GET /api/documents?archived=true

---

## 9. Duplicar Documento

### 9.1. Duplicar

**Endpoint**: POST /api/documents/:id/duplicate

**Comportamento:**
1. Cria cÃ³pia do documento
2. TÃ­tulo: "[CÃ³pia] TÃ­tulo Original"
3. Copia todo o conteÃºdo
4. **NÃƒO copia** sub-pÃ¡ginas (apenas documento raiz)
5. **NÃƒO copia** compartilhamentos
6. Respeita limites do plano

**OpÃ§Ã£o de Duplicar com Sub-pÃ¡ginas:**

**Payload:**
```json
{
  "includeSubPages": true
}
```

Se `true`, duplica recursivamente todas sub-pÃ¡ginas.

---

## 10. Busca

### 10.1. Busca Global

**Endpoint**: GET /api/documents/search

**Query Params:**
- `q` (string, obrigatÃ³rio) - Termo de busca
- `in` (title|content|both, default: both)
- `page`, `size`, `sort`

**Exemplo:**
```
GET /api/documents/search?q=projeto&in=title
```

**Busca em:**
- TÃ­tulo do documento
- ConteÃºdo do documento (full-text search)
- Tags (futuro)

**Resposta:**
```json
{
  "content": [
    {
      "id": "uuid",
      "title": "Projeto X",
      "excerpt": "...texto com <mark>projeto</mark> destacado...",
      "match": "title",
      "updatedAt": "2024-01-01T00:00:00Z"
    }
  ],
  "totalElements": 12
}
```

### 10.2. Full-Text Search (PostgreSQL)

**Ãndice:**
```sql
CREATE INDEX idx_documents_content_search 
ON documents 
USING gin(to_tsvector('portuguese', content));
```

**Query:**
```kotlin
@Query("""
    SELECT d FROM Document d 
    WHERE d.owner = :owner 
    AND (
        to_tsvector('portuguese', d.content) @@ plainto_tsquery('portuguese', :query)
        OR LOWER(d.title) LIKE LOWER(CONCAT('%', :query, '%'))
    )
""")
fun searchDocuments(owner: User, query: String): List<Document>
```

---

## 11. HistÃ³rico de VersÃµes (Futuro - Premium)

### 11.1. Snapshot de VersÃ£o

```kotlin
@Entity
@Table(name = "document_versions")
data class DocumentVersion(
    @Id
    val id: UUID = UUID.randomUUID(),
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "document_id")
    val document: Document,
    
    @Column(columnDefinition = "TEXT")
    val content: String,
    
    @Column
    val version: Int,
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "created_by")
    val createdBy: User,
    
    @CreatedDate
    val createdAt: LocalDateTime = LocalDateTime.now()
)
```

### 11.2. EstratÃ©gia de Versionamento

**Criar snapshot:**
- A cada 10 salvamentos
- A cada hora (se houver mudanÃ§as)
- Manual (usuÃ¡rio clica "Salvar versÃ£o")

**Manter:**
- Ãšltimas 50 versÃµes
- VersÃµes manuais: sempre

### 11.3. Restaurar VersÃ£o

**Endpoint**: POST /api/documents/:id/versions/:versionId/restore

**Comportamento:**
1. Cria novo snapshot da versÃ£o atual
2. Restaura conteÃºdo da versÃ£o especificada
3. Incrementa nÃºmero de versÃ£o

---

## 12. Mensagens de Erro

```json
{
  "error": {
    "code": "DOCUMENT_NOT_FOUND",
    "message": "Documento nÃ£o encontrado"
  }
}
```

**CÃ³digos de Erro:**
- `DOCUMENT_NOT_FOUND` - Documento nÃ£o existe
- `ACCESS_DENIED` - Sem permissÃ£o
- `INVALID_PARENT` - PÃ¡gina pai invÃ¡lida
- `MAX_DEPTH_EXCEEDED` - MÃ¡ximo 2 nÃ­veis de hierarquia
- `IMAGE_TOO_LARGE` - Imagem excede 5MB
- `INVALID_IMAGE_FORMAT` - Formato nÃ£o suportado
- `VERSION_CONFLICT` - Conflito de versÃ£o (salvamento simultÃ¢neo)
- `STORAGE_LIMIT_EXCEEDED` - Limite de storage atingido
