# Regras de Neg√≥cio - Compartilhamento e Colabora√ß√£o

## 1. Compartilhamento de Documentos (Premium)

### 1.1. Vis√£o Geral

**Requisito:** Plano Premium

**Funcionalidades:**
- Compartilhar documentos com outros usu√°rios
- Definir permiss√µes (visualizar ou editar)
- Gerenciar acessos
- Link p√∫blico (opcional)

### 1.2. Model de Compartilhamento

```kotlin
@Entity
@Table(name = "document_shares")
data class DocumentShare(
    @Id
    val id: UUID = UUID.randomUUID(),
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "document_id", nullable = false)
    val document: Document,
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "shared_by", nullable = false)
    val sharedBy: User,
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "shared_with", nullable = false)
    val sharedWith: User,
    
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    val permission: SharePermission,
    
    @CreatedDate
    val createdAt: LocalDateTime = LocalDateTime.now(),
    
    @Column
    val expiresAt: LocalDateTime? = null,
    
    @Column(nullable = false)
    val isActive: Boolean = true
)

enum class SharePermission {
    VIEW,   // Apenas visualizar
    EDIT    // Visualizar e editar
}
```

### 1.3. Compartilhar Documento

**Endpoint**: POST /api/documents/:id/share

**Payload:**
```json
{
  "shareWith": [
    {
      "userIdentifier": "joao@example.com",  // Email ou username
      "permission": "EDIT"
    },
    {
      "userIdentifier": "maria",
      "permission": "VIEW"
    }
  ],
  "message": "Confira este documento!"  // Opcional
}
```

**Valida√ß√µes:**
1. Documento existe?
2. Usu√°rio √© o dono?
3. Plano do dono √© Premium?
4. Usu√°rios a compartilhar existem?
5. N√£o compartilhar com si mesmo
6. N√£o duplicar compartilhamentos

**Resposta (HTTP 200):**
```json
{
  "shares": [
    {
      "id": "uuid",
      "user": {
        "id": "uuid",
        "name": "Jo√£o Silva",
        "username": "joao",
        "email": "joao@example.com",
        "avatar": "https://..."
      },
      "permission": "EDIT",
      "createdAt": "2024-01-01T12:00:00Z"
    }
  ],
  "notFound": ["usuario_inexistente@email.com"]
}
```

**Enviar Email de Notifica√ß√£o:**
```
Assunto: [Nome do Dono] compartilhou um documento com voc√™

Ol√° [Nome],

[Nome do Dono] compartilhou o documento "[T√≠tulo]" com voc√™.

[Visualizar Documento]

Mensagem: "Confira este documento!"

Atenciosamente,
Notion Clone
```

### 1.4. Listar Compartilhamentos

**Endpoint**: GET /api/documents/:id/shares

**Resposta:**
```json
{
  "shares": [
    {
      "id": "uuid",
      "user": {
        "id": "uuid",
        "name": "Jo√£o Silva",
        "username": "joao",
        "avatar": "https://..."
      },
      "permission": "EDIT",
      "createdAt": "2024-01-01T12:00:00Z"
    }
  ]
}
```

### 1.5. Atualizar Permiss√£o

**Endpoint**: PATCH /api/documents/:docId/shares/:shareId

**Payload:**
```json
{
  "permission": "VIEW"
}
```

**Valida√ß√µes:**
1. Usu√°rio √© o dono do documento?
2. Compartilhamento existe?

### 1.6. Revogar Acesso

**Endpoint**: DELETE /api/documents/:docId/shares/:shareId

**Valida√ß√µes:**
1. Usu√°rio √© o dono do documento?
2. Compartilhamento existe?

**Comportamento:**
- Atualiza `isActive` para `false`
- Emite evento WebSocket se usu√°rio estiver online
- Frontend fecha documento automaticamente

**Enviar Email:**
```
Assunto: Acesso revogado ao documento "[T√≠tulo]"

Ol√° [Nome],

[Nome do Dono] revogou seu acesso ao documento "[T√≠tulo]".

Se tiver d√∫vidas, entre em contato com o dono do documento.

Atenciosamente,
Notion Clone
```

### 1.7. Documentos Compartilhados Comigo

**Endpoint**: GET /api/documents/shared-with-me

**Query Params:**
- `permission` (VIEW|EDIT, opcional)
- `page`, `size`, `sort`

**Resposta:**
```json
{
  "content": [
    {
      "id": "uuid",
      "title": "Documento Compartilhado",
      "icon": "üìÑ",
      "owner": {
        "name": "Jo√£o Silva",
        "username": "joao",
        "avatar": "https://..."
      },
      "permission": "EDIT",
      "sharedAt": "2024-01-01T12:00:00Z",
      "updatedAt": "2024-01-02T15:30:00Z"
    }
  ],
  "totalElements": 5
}
```

---

## 2. Link P√∫blico (Premium)

### 2.1. Gerar Link P√∫blico

**Endpoint**: POST /api/documents/:id/public-link

**Payload:**
```json
{
  "permission": "VIEW",  // VIEW ou EDIT
  "expiresAt": "2024-12-31T23:59:59Z",  // Opcional
  "password": "senha123"  // Opcional
}
```

**Comportamento:**
1. Gera token √∫nico e seguro (UUID ou hash)
2. Link: `https://notion-clone.com/public/[token]`
3. Salva no banco com configura√ß√µes

**Model:**
```kotlin
@Entity
@Table(name = "public_links")
data class PublicLink(
    @Id
    val id: UUID = UUID.randomUUID(),
    
    @Column(unique = true, nullable = false)
    val token: String = generateSecureToken(),
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "document_id", nullable = false)
    val document: Document,
    
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    val permission: SharePermission,
    
    @Column
    val password: String? = null,  // Hashed
    
    @Column
    val expiresAt: LocalDateTime? = null,
    
    @Column(nullable = false)
    val isActive: Boolean = true,
    
    @Column(nullable = false)
    val viewCount: Int = 0,
    
    @CreatedDate
    val createdAt: LocalDateTime = LocalDateTime.now()
)
```

**Resposta:**
```json
{
  "id": "uuid",
  "url": "https://notion-clone.com/public/secure-token-123",
  "permission": "VIEW",
  "expiresAt": "2024-12-31T23:59:59Z",
  "hasPassword": true,
  "createdAt": "2024-01-01T12:00:00Z"
}
```

### 2.2. Acessar Link P√∫blico

**Endpoint**: GET /public/:token

**Valida√ß√µes:**
1. Token existe?
2. Link est√° ativo?
3. Link n√£o expirou?
4. Se tem senha, validar senha

**Fluxo com Senha:**
```
1. Usu√°rio acessa /public/token
2. Se hasPassword = true, mostrar modal de senha
3. POST /public/:token/verify-password { password: "..." }
4. Se correto, gerar session token tempor√°rio
5. Redirecionar para visualiza√ß√£o do documento
```

**Resposta (sem senha):**
- Retorna documento completo em modo visualiza√ß√£o
- Desabilita edi√ß√£o se permission = VIEW
- Incrementa viewCount

**Resposta (com senha - sem autenticar):**
```json
{
  "requiresPassword": true,
  "documentTitle": "Documento P√∫blico",
  "owner": "Jo√£o Silva"
}
```

### 2.3. Listar Links P√∫blicos

**Endpoint**: GET /api/documents/:id/public-links

**Resposta:**
```json
{
  "links": [
    {
      "id": "uuid",
      "url": "https://notion-clone.com/public/token",
      "permission": "VIEW",
      "hasPassword": true,
      "expiresAt": "2024-12-31T23:59:59Z",
      "viewCount": 42,
      "createdAt": "2024-01-01T12:00:00Z"
    }
  ]
}
```

### 2.4. Desabilitar Link P√∫blico

**Endpoint**: DELETE /api/documents/:docId/public-links/:linkId

**Comportamento:**
- Atualiza `isActive` para `false`
- Link para de funcionar imediatamente

---

## 3. Colabora√ß√£o em Tempo Real (Premium)

### 3.1. Vis√£o Geral

**Requisito:** Plano Premium do DONO do documento

**Funcionalidades:**
- Edi√ß√£o simult√¢nea
- Ver cursores de outros usu√°rios
- Ver quem est√° online
- Sincroniza√ß√£o em tempo real
- Presen√ßa de usu√°rios

### 3.2. Tecnologia

**Protocolo:** WebSocket (STOMP sobre WebSocket)

**Endpoints WebSocket:**
```
/ws/connect           ‚Üí Conectar ao WebSocket
/topic/document/{id}  ‚Üí Subscrever ao documento
/app/document/{id}    ‚Üí Enviar mudan√ßas
```

### 3.3. Conectar ao Documento

**Cliente (Frontend):**
```typescript
const socket = new SockJS('http://localhost:8080/ws/connect');
const stompClient = Stomp.over(socket);

stompClient.connect({ 
  Authorization: `Bearer ${accessToken}` 
}, () => {
  // Subscrever ao t√≥pico do documento
  stompClient.subscribe(`/topic/document/${docId}`, (message) => {
    const event = JSON.parse(message.body);
    handleCollaborationEvent(event);
  });
  
  // Anunciar presen√ßa
  stompClient.send(`/app/document/${docId}/join`, {}, JSON.stringify({
    userId: currentUser.id,
    username: currentUser.username,
    avatar: currentUser.avatar
  }));
});
```

**Servidor (Backend):**
```kotlin
@MessageMapping("/document/{id}/join")
@SendTo("/topic/document/{id}")
fun joinDocument(
    @DestinationVariable id: UUID,
    @Payload joinMessage: JoinMessage,
    principal: Principal
): CollaborationEvent {
    // Validar permiss√µes
    validateCanCollaborate(id, principal)
    
    // Adicionar √† lista de presen√ßa
    presenceService.addUser(id, joinMessage.userId)
    
    // Broadcast para todos
    return CollaborationEvent(
        type = EventType.USER_JOINED,
        documentId = id,
        user = joinMessage,
        timestamp = Instant.now()
    )
}
```

### 3.4. Tipos de Eventos

```kotlin
enum class CollaborationEventType {
    USER_JOINED,      // Usu√°rio entrou
    USER_LEFT,        // Usu√°rio saiu
    CONTENT_CHANGED,  // Conte√∫do mudou
    CURSOR_MOVED,     // Cursor de usu√°rio moveu
    SELECTION_CHANGED // Sele√ß√£o de texto mudou
}

data class CollaborationEvent(
    val type: CollaborationEventType,
    val documentId: UUID,
    val user: UserPresence,
    val data: Any? = null,
    val timestamp: Instant
)

data class UserPresence(
    val userId: UUID,
    val username: String,
    val avatar: String?,
    val color: String  // Cor √∫nica para o usu√°rio
)
```

### 3.5. Sincroniza√ß√£o de Conte√∫do

**Tiptap Collaboration Extension:**

Usa **Y.js** para CRDTs (Conflict-free Replicated Data Types):

```typescript
import { TiptapCollabProvider } from '@hocuspocus/provider'

const provider = new TiptapCollabProvider({
  name: documentId,
  appId: 'notion-clone',
  token: accessToken,
  websocketProvider: stompClient
})

const editor = new Editor({
  extensions: [
    StarterKit,
    Collaboration.configure({
      document: provider.document,
    }),
    CollaborationCursor.configure({
      provider: provider,
      user: {
        name: currentUser.username,
        color: getUserColor(currentUser.id)
      }
    })
  ]
})
```

**Operational Transformation:**
1. Usu√°rio A digita "Hello"
2. Frontend gera opera√ß√£o: `insert("Hello", position: 0)`
3. Envia via WebSocket
4. Backend recebe e valida
5. Backend aplica transforma√ß√£o (Y.js)
6. Backend broadcast para todos os clientes
7. Clientes aplicam opera√ß√£o localmente

**Tratamento de Conflitos:**
- Y.js gerencia automaticamente conflitos
- CRDTs garantem converg√™ncia eventual
- Todos usu√°rios chegam ao mesmo estado final

### 3.6. Presen√ßa de Usu√°rios

**Endpoint**: GET /api/documents/:id/presence

**Resposta:**
```json
{
  "users": [
    {
      "userId": "uuid",
      "username": "joao",
      "avatar": "https://...",
      "color": "#FF5733",
      "cursor": {
        "position": 123,
        "timestamp": "2024-01-01T12:30:45Z"
      }
    },
    {
      "userId": "uuid2",
      "username": "maria",
      "avatar": "https://...",
      "color": "#33FF57",
      "cursor": {
        "position": 456,
        "timestamp": "2024-01-01T12:30:50Z"
      }
    }
  ],
  "count": 2
}
```

**Atualiza√ß√£o em Tempo Real:**
- WebSocket envia eventos de presen√ßa
- Frontend atualiza UI mostrando avatares dos usu√°rios online
- Mostra cursores de outros usu√°rios no editor

**UI de Presen√ßa:**
```
‚îå‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îê
‚îÇ Meu Documento                  üë§ üë§    ‚îÇ  ‚Üê Avatares de quem est√° online
‚îú‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î§
‚îÇ # T√≠tulo                                ‚îÇ
‚îÇ                                         ‚îÇ
‚îÇ Par√°grafo de texto...‚ñà                  ‚îÇ  ‚Üê Cursor do Jo√£o
‚îÇ                      [Jo√£o]             ‚îÇ
‚îÇ                                         ‚îÇ
‚îÇ Mais texto aqui...      ‚ñà               ‚îÇ  ‚Üê Cursor da Maria
‚îÇ                        [Maria]          ‚îÇ
‚îî‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îò
```

### 3.7. Desconex√£o

**Detectar Desconex√£o:**
- Cliente envia heartbeat a cada 30 segundos
- Se n√£o receber heartbeat por 1 minuto, considerar desconectado
- Remover da lista de presen√ßa
- Broadcast evento USER_LEFT

**Reconex√£o:**
- Cliente tenta reconectar automaticamente
- Sincroniza estado atual do documento
- Reanuncia presen√ßa

### 3.8. Limita√ß√µes

**M√°ximo de Usu√°rios Simult√¢neos:**
- FREE: 0 (n√£o suporta colabora√ß√£o)
- PRO: 0 (n√£o suporta colabora√ß√£o)
- PREMIUM: 10 usu√°rios simult√¢neos por documento

**Se exceder:**
- Novos usu√°rios entram em "modo visualiza√ß√£o apenas"
- Mostrar mensagem: "Limite de colaboradores atingido. Voc√™ est√° em modo visualiza√ß√£o."

---

## 4. Coment√°rios (Futuro)

### 4.1. Adicionar Coment√°rio

**Endpoint**: POST /api/documents/:id/comments

**Payload:**
```json
{
  "content": "√ìtimo ponto!",
  "position": {
    "blockId": "uuid",
    "start": 10,
    "end": 20
  }
}
```

**Model:**
```kotlin
@Entity
@Table(name = "document_comments")
data class DocumentComment(
    @Id
    val id: UUID = UUID.randomUUID(),
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "document_id", nullable = false)
    val document: Document,
    
    @Column(columnDefinition = "TEXT", nullable = false)
    val content: String,
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "author_id", nullable = false)
    val author: User,
    
    @Column
    val blockId: UUID?,  // Bloco onde foi comentado
    
    @Column
    val startPosition: Int?,
    
    @Column
    val endPosition: Int?,
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent_id")
    val parent: DocumentComment? = null,  // Para respostas
    
    @Column(nullable = false)
    val isResolved: Boolean = false,
    
    @CreatedDate
    val createdAt: LocalDateTime = LocalDateTime.now(),
    
    @LastModifiedDate
    val updatedAt: LocalDateTime = LocalDateTime.now()
)
```

### 4.2. Listar Coment√°rios

**Endpoint**: GET /api/documents/:id/comments

**Query Params:**
- `resolved` (boolean, opcional)
- `blockId` (UUID, opcional) - Coment√°rios de um bloco espec√≠fico

### 4.3. Resolver Coment√°rio

**Endpoint**: PATCH /api/documents/:docId/comments/:commentId/resolve

**Comportamento:**
- Atualiza `isResolved` para `true`
- Envia notifica√ß√£o para autor original

---

## 5. Notifica√ß√µes

### 5.1. Tipos de Notifica√ß√µes

```kotlin
enum class NotificationType {
    DOCUMENT_SHARED,      // Documento compartilhado com voc√™
    SHARE_PERMISSION_CHANGED,  // Sua permiss√£o mudou
    ACCESS_REVOKED,       // Seu acesso foi revogado
    COMMENT_ADDED,        // Algu√©m comentou
    COMMENT_REPLIED,      // Algu√©m respondeu seu coment√°rio
    COMMENT_RESOLVED,     // Coment√°rio foi resolvido
    MENTION,              // Voc√™ foi mencionado (@usuario)
    DOCUMENT_UPDATED      // Documento compartilhado foi atualizado
}
```

### 5.2. Model de Notifica√ß√£o

```kotlin
@Entity
@Table(name = "notifications")
data class Notification(
    @Id
    val id: UUID = UUID.randomUUID(),
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    val user: User,
    
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    val type: NotificationType,
    
    @Column(nullable = false)
    val title: String,
    
    @Column(columnDefinition = "TEXT")
    val message: String,
    
    @Column
    val relatedDocumentId: UUID?,
    
    @Column
    val relatedUserId: UUID?,
    
    @Column(nullable = false)
    val isRead: Boolean = false,
    
    @CreatedDate
    val createdAt: LocalDateTime = LocalDateTime.now()
)
```

### 5.3. Listar Notifica√ß√µes

**Endpoint**: GET /api/notifications

**Query Params:**
- `read` (boolean, opcional)
- `type` (NotificationType, opcional)
- `page`, `size`

**Resposta:**
```json
{
  "content": [
    {
      "id": "uuid",
      "type": "DOCUMENT_SHARED",
      "title": "Novo documento compartilhado",
      "message": "Jo√£o Silva compartilhou 'Projeto X' com voc√™",
      "relatedDocument": {
        "id": "uuid",
        "title": "Projeto X"
      },
      "relatedUser": {
        "id": "uuid",
        "name": "Jo√£o Silva",
        "username": "joao"
      },
      "isRead": false,
      "createdAt": "2024-01-01T12:00:00Z"
    }
  ],
  "unreadCount": 5,
  "totalElements": 42
}
```

### 5.4. Marcar como Lida

**Endpoint**: PATCH /api/notifications/:id/read

**Endpoint**: PATCH /api/notifications/read-all

### 5.5. WebSocket para Notifica√ß√µes em Tempo Real

**Cliente subscreve:**
```
/user/queue/notifications
```

**Servidor envia:**
```json
{
  "type": "NEW_NOTIFICATION",
  "notification": {
    "id": "uuid",
    "type": "DOCUMENT_SHARED",
    "title": "...",
    "message": "..."
  }
}
```

**Frontend:**
- Mostra badge com contador de n√£o lidas
- Toast notification
- Som (opcional, nas prefer√™ncias)

---

## 6. Auditoria de Acesso

### 6.1. Log de Acessos

```kotlin
@Entity
@Table(name = "document_access_log")
data class DocumentAccessLog(
    @Id
    val id: UUID = UUID.randomUUID(),
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "document_id", nullable = false)
    val document: Document,
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id")
    val user: User?,  // null se acesso p√∫blico
    
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    val action: AccessAction,
    
    @Column
    val ipAddress: String,
    
    @Column
    val userAgent: String,
    
    @CreatedDate
    val timestamp: LocalDateTime = LocalDateTime.now()
)

enum class AccessAction {
    VIEW,
    EDIT,
    SHARE,
    DOWNLOAD_PDF,
    PUBLIC_LINK_ACCESS
}
```

### 6.2. Endpoint de Auditoria

**Endpoint**: GET /api/documents/:id/access-log

**Resposta:**
```json
{
  "content": [
    {
      "user": {
        "name": "Jo√£o Silva",
        "username": "joao"
      },
      "action": "EDIT",
      "timestamp": "2024-01-01T12:30:00Z",
      "ipAddress": "192.168.1.1"
    },
    {
      "user": null,
      "action": "PUBLIC_LINK_ACCESS",
      "timestamp": "2024-01-01T13:00:00Z",
      "ipAddress": "203.0.113.0"
    }
  ]
}
```

---

## 7. Mensagens de Erro

```json
{
  "error": {
    "code": "COLLABORATION_NOT_AVAILABLE",
    "message": "Colabora√ß√£o em tempo real requer plano Premium",
    "details": {
      "currentPlan": "PRO",
      "requiredPlan": "PREMIUM"
    }
  }
}
```

**C√≥digos de Erro:**
- `COLLABORATION_NOT_AVAILABLE` - Plano insuficiente
- `MAX_COLLABORATORS_REACHED` - Limite de colaboradores atingido
- `SHARE_NOT_ALLOWED` - N√£o pode compartilhar (plano insuficiente)
- `INVALID_PERMISSION` - Permiss√£o inv√°lida
- `USER_NOT_FOUND` - Usu√°rio para compartilhar n√£o existe
- `ALREADY_SHARED` - J√° compartilhado com este usu√°rio
- `LINK_EXPIRED` - Link p√∫blico expirado
- `INVALID_PASSWORD` - Senha do link p√∫blico incorreta
